numpy/core/src/multiarray/ctors.c

```diff
@@ -2874,17 +2936,54 @@ PyArray_Zeros(int nd, npy_intp const *dims, PyArray_Descr *type, int is_f_order)
  */
 NPY_NO_EXPORT PyObject *
 PyArray_Empty(int nd, npy_intp const *dims, PyArray_Descr *type, int is_f_order)
+{
+    npy_dtype_info dt_info = {NULL, NULL};
+
+    int res = PyArray_ExtractDTypeAndDescriptor(
+        type, &dt_info.descr, &dt_info.dtype);
+
+    // steal reference
+    Py_XDECREF(type);
+
+    if (res < 0) {
+        Py_XDECREF(dt_info.descr);
+        Py_XDECREF(dt_info.dtype);
```

Comment:
```suggestion
```
I think `PyArray_ExtractDTypeAndDescriptor` cleans up on error (and I think that is much easier to reason about if functions guarantee to do nothing on errors, although there are exceptions like the argument parsing, so TBH, maybe it doesn't matter).

Reply:
It looks like `PyArray_ExtractDTypeAndDescriptor can't ever fail right now, so it doesn't matter much. Changed it to reflect your expectations so in the future if it can fail the references will be cleaned up correctly.